### PR TITLE
New contributors page

### PR DESCRIPTION
**Pull request by Shellyber**
## Status
n Progress

## Related Issues
fixes: https://github.com/demisto/etc/issues/37815

## Description
Added new Contributors page

## Screenshots
Paste here any images that will help the reviewer

### CODE REVIEW

**Comment by Shellyber** [comment]
@Itay4 Please review

**Comment by Itay4** [comment]
currently each column in the grid has different width, please align all to be of the same width

File: content-repo/gen_top_contrib.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            if not user == 'xsoar-bot':
+                users.append({
+                    'Contributor': f"<img src='{response.get('user').get('avatar_url')}'/><br></br> "
+                                   f"<a href='{github_profile}' target='_blank'>{user}</a>"
+                })
+
+            if user == 'xsoar-bot':
+                pr_body = response.get('body')
+                if 'Contributor' in pr_body:
+                    contributor = re.search(r"(?<=@)[a-zA-z-0-9]+", pr_body)[0].replace('\n', '')
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:172]: better to compile the regex once and then use it as done for `PR_NUMBER_REGEX`

File: content-repo/gen_top_contrib.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+    github_avatar = response.get('avatar_url')
+    github_profile = response.get('html_url')
+
+    return github_avatar, github_profile
+
+
+def get_pr_user():
+    users = []
+    _, inner_prs = get_contrib_prs()
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:144]: why are we returning `pr_bodies` from `get_contrib_prs` if we're not using it?

File: content-repo/gen_top_contrib.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    res = requests.request('GET', url, headers=HEADERS, verify=False)
+    response = res.json()
+
+    github_avatar = response.get('avatar_url')
+    github_profile = response.get('html_url')
+
+    return github_avatar, github_profile
+
+
+def get_pr_user():
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:142]: lets split this function into 2: one to retrieve the users and one to do the manipulation on the result of the http response

File: content-repo/gen_top_contrib.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        user.update({'Number of Contributions': prs})
+
+    list_users = []
+    result = {i['Contributor']: i for i in reversed(users)}.values()
+    new_res = sorted(result, key=lambda k: k['Number of Contributions'], reverse=True)
+
+    for user in new_res:
+        user['Contributor'] += f'<br></br>{user["Number of Contributions"]} Contributions'
+        list_users.append(user['Contributor'])
+    res = create_grid(list_users)
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:189]: call `create_grid` from `main`

File: content-repo/gen_top_contrib.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    parser = argparse.ArgumentParser(description='Generate Top contributors page.',
+                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument("-t", "--target", help="Target dir to generate docs at.", required=True)
+    args = parser.parse_args()
+    start_mdx_server()
+    contrib_base = f'{os.path.dirname(os.path.abspath(__file__))}/top-contributors.md'
+    if not os.path.exists(contrib_base):
+        os.makedirs(contrib_base)
+    contrib_target = args.target + '/top-contributors.md'
+    shutil.copy(contrib_base, contrib_target)
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:204]: not sure i follow the logic here, why are we copying here ?

File: content-repo/gen_top_contrib.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+    return res
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Generate Top contributors page.',
+                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument("-t", "--target", help="Target dir to generate docs at.", required=True)
+    args = parser.parse_args()
+    start_mdx_server()
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:199]: why do we need mdx server here?

File: content-repo/gen_top_contrib.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    shutil.copy(contrib_base, contrib_target)
+    top_table = get_pr_user()
+    with open(contrib_target, 'a', encoding='utf-8') as f:
+        f.write(f'\n {top_table}')
+
+    print('Stopping mdx server ...')
+    stop_mdx_server()
+
+
+if __name__ == "__main__":
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:213]:
```suggestion
if __name__ == '__main__':
```

File: content-repo/top-contributors.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+---
+id: top-contributors
+title: Top XSOAR Contributors
+sidebar_label: Index
+custom_edit_url: ""
+slug: /marketplace/contributors
```
Comment (Itay4) [inline content-repo/top-contributors.md:6]: can you please explain why we need those 3?

File: src/pages/marketplace/contributors.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import React from 'react'
+import Contributors from '../marketplace/top-contributors.md'
+import Layout from '@theme/Layout';
+import styles from "./styles.module.css";
+
+
+function topContrib() {
```
Comment (Itay4) [inline src/pages/marketplace/contributors.js:7]: why is this needed?

**Review by Itay4 — CHANGES_REQUESTED** [review]
looks good 👍 
1. please add unit tests
2. please use `raise_for_status()` for all the requests you're sending

File: content-repo/gen_top_contrib.py (hunk 2 above)
Comment (Shellyber) [reply content-repo/gen_top_contrib.py:144]: We will use it at the future. 
once [this pr](https://github.com/demisto/etc/issues/38530) will be ready. I won't have to make the get_pr_user for new contributors. I'll be able to take the username directly from the pr body.

File: src/pages/marketplace/contributors.js (hunk 9 above)
Comment (Shellyber) [reply src/pages/marketplace/contributors.js:7]: So I can't put style easily on the markdown file itself. It better to handle this in a js file where I can control whatever I need and want

File: content-repo/gen_top_contrib.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+import re
+import shutil
+
+import requests
+
+# Disable insecure warnings
+import urllib3
+from mdx_utils import (start_mdx_server, stop_mdx_server)
+
+urllib3.disable_warnings()
```
[the remark this answers is not in the document]
Comment (Itay4) [reply content-repo/gen_top_contrib.py:11]: note my comment here, we should run requests with `verify=False`

File: content-repo/gen_top_contrib.py [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    """
+    Get the github user.
+    Args:
+        user_name (str): the github username.
+
+    Returns: The user avatar and its profile.
+
+    """
+    url = f'{URL}/users/{user_name}'
+    res = requests.request('GET', url, headers=HEADERS, verify=False)
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:154]: missing `raise_for_status`

File: content-repo/gen_top_contrib.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+    Returns: prs (dict)
+    """
+    prs = []
+    if link := res.headers.get('Link'):
+        links = link.split(',')
+        for link in links:
+            last_page = link[link.find("<")+1:link.find(">")][-1]
+        while params['page'] <= int(last_page):
+            params['page'] = params['page'] + 1
+            response = requests.request('GET', url, params=params, headers=HEADERS, verify=False)
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:83]: missing `raise_for_status`

File: content-repo/get_top_contrib_test.py [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+            "body": "## Original External PR\r\n[external pull request](https://github.com/demisto/content/pull/13614)"
+                    "\r\n\r\n## Contributing to Cortex XSOAR Content",
+        }
+    ]
+
+    _, res = get_external_prs(mock_response)
+    assert ['13801', '13614'] == res
+
+
+def test_get_github_user(requests_mock):
```
Comment (Itay4) [inline content-repo/get_top_contrib_test.py:370]: please make sure all tests have docstrings

File: content-repo/gen_top_contrib.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+    for i in range(0, len(dataset), row_length):
+        html_card += '<tr>'
+        for element in dataset[i:i+row_length]:
+            html_card += f'\n<td>{element} </td>\n'
+        html_card += '</tr>\n'
+
+    return html_card
+
+
+def get_external_prs(prs: list):
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:44]: please add return type hint

File: content-repo/gen_top_contrib.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+
+    next_pages_prs = github_pagination_prs(url, params, res)
+    prs.extend(next_pages_prs)
+
+    _, inner_prs = get_external_prs(prs)
+
+    return inner_prs
+
+
+def get_contrib_prs():
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:115]: please add return type hint

File: content-repo/gen_top_contrib.py [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+    pr_bodies, inner_prs = get_external_prs(prs)
+    contractors_prs = get_contractors_prs()
+    for pr in inner_prs:
+        if pr in contractors_prs:
+            inner_prs.remove(pr)
+
+    return pr_bodies, inner_prs
+
+
+def get_github_user(user_name: str):
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:144]: please add return type hint

File: src/pages/marketplace/contributors.js [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+  return (
+    <Layout title="Contributors">
+    <h1 className={styles.h1Contrib}>Top Contributors</h1>
```
Comment (Itay4) [inline src/pages/marketplace/contributors.js:10]: please lint properly and add indentation

**Review by Itay4 — COMMENTED** [review]
please note the guardrails and lgtm alerts

File: content-repo/gen_top_contrib.py (hunk 11 above)
Comment (Shellyber) [reply content-repo/gen_top_contrib.py:154]: I don't want to use here raise for status. 
I have a case where user was not found, and getting 404.
I'm handling this with an `if` statement in `get_contributors_users` command.
Don't want this function to fail

File: content-repo/gen_top_contrib.py [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+import os
+import re
+
+import requests
+import urllib3
+
+# Disable insecure warnings
+urllib3.disable_warnings()
+
+
```
Comment (Itay4) [inline content-repo/gen_top_contrib.py:13]: please override `print` with `timestamped_print` as we do in the other python scripts

Comment (Shellyber) [reply content-repo/gen_top_contrib.py:9]: Don't think it is necessary here, we don't care about the timestamps here. 
We use the print only for 404 not found. the time here is not important.

Comment (Itay4) [reply content-repo/gen_top_contrib.py:9]: it helps when debugging in the build console, lets add it

Comment (Shellyber) [reply content-repo/gen_top_contrib.py:10]: Done

**Review by Itay4 — APPROVED** [review]
please fix the lgtm alerts before merging
